rgw: implement S3 additional checksum support

mattbenjamin opened the pull request:
Adds new Blake3 digest format (native), a concrete type to
represent digests, and static_visitor machinery to unify varying
checksum computations.

This framework, together with new trailing checksum header support,
is used to implement S3 additional checksum verification.  Parts of
the AWS content checksum API work build on a prior contribution from
imtzw <tongzhiwei_yewu@cmss.chinamobile.com>.
Thank you!

Fixes: https://tracker.ceph.com/issues/42080
Fixes: https://tracker.ceph.com/issues/63951

squashed commits:
* rgw_cksum: add trival test vectors for sha-format digests
      Computed digests match those produced by sha1sum, sha256sum,
      and sha512sum utilities.
* rgw_cksum: add test vectors for blake3
      Tests the same input strings with digests validated by
      b3sum (https://crates.io/crates/b3sum).
* rgw_ckum: switch to accel crc32c
      The internal Ceph convention appears to be to omit a final
      xor where ceph_crc32c is used, but it's required for compatibility
      with AWS implementations.
* rgw_cksum: add XXH3 digest
* rgw_cksum: write class encoder for rgw::digest::Cksum
* rgw_cksum: also reverse crc32c (REBASEME)
      Mark noticed that the crc32c output was being tested against a
      byteswapped value (crc32c also needs byteswap on LE).
* rgw_cksum: add digest::Cksum serde tests
* rgw_cksum: fix main(...) linkage
      (so we run our main unit and not the one in gmock
* rgw_cksum: convenience extensions for integration with RGW/S3
* introduce rgw_cksum unique_ptr factory
* rgw_cksum: mark string transform accessors const
* rgw_cksum: fixup unittest_rgw_chksum compilation--all existing tests pass
* rgw_cksum: hook up put-object checksum workflow
* tweaks to report on content checksum mismatch
* rgw_cksum: match SDK as well as general checksum header
* make it more efficient
* initialize RGWPutObj_Cksum::digest
* rgw_cksum:  write parse_cksum_type w/const char* arg
* initialize _type correctly; doing armored wrong
* fix expected checksum header name, clean up verify
* fix output on checksum verify fail, cleanup
* introduce Cksum::to_armor();  all AWS cases pass
* oops, extra 0-byte at end of to_armor() result

Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

mattbenjamin commented on src/rgw/rgw_cksum.h:
```diff
@@ -0,0 +1,296 @@
+      std::string hs;
+      const auto& ckd = checksums[uint16_t(type)];
+      hs.resize(ckd.armored_size);
+      memset(hs.data(), 0, hs.length());
```
almost certainly can remove memset;  I introduced it for debugging

mattbenjamin commented on src/rgw/rgw_cksum.h:
```diff
@@ -0,0 +1,296 @@
+      return hs;
+    }
+
+    std::string to_base64() const {
```
should add comment:  this encoding is here because it is the one used by all of the command-line checksum tools we found (md5sum, sha1sum, sha256sum, bsum, &c)

mattbenjamin commented on src/rgw/rgw_cksum.h:
```diff
@@ -0,0 +1,296 @@
+      return rgw::to_base64(hex());
+    }
+
+    std::string to_string() const  {
```
to_string doesn't reflect any existing practice, I based it (sort of) on self-describing X.500 encodings;  the idea was to create a checksum value that included the encoding type as text;  Since AWS always uses a header, there's maybe no need for it

mattbenjamin commented on src/rgw/rgw_basic_types.h:
```diff
@@ -255,6 +256,7 @@ struct RGWUploadPartInfo {
   uint64_t size;
   uint64_t accounted_size{0};
   std::string etag;
+  rgw::cksum::Cksum cksum;
```
looks like I did update RGWUploadPartInfo, following imtz's lead.  I kind of favor this, even if we don't immediately implement a consistency check?

cbodley replied:
i think we need this to calculate the checksums for multipart uploads in `RadosMultipartUpload::complete()`. that has access to the list of `RGWUploadPartInfo`s, but doesn't read each part object for xattrs like RGW_ATTR_AMZ_CKSUM

mattbenjamin commented on src/rgw/rgw_cksum_pipe.cc:
```diff
@@ -0,0 +1,84 @@
+	  std::move(algo_header));
+    }
+    /* XXXX this is terminating radosgw */
+    throw rgw::io::Exception(ERR_BAD_DIGEST, std::system_category());
```
need to do something about this

mattbenjamin replied:
will look at it

mattbenjamin commented on src/rgw/rgw_common.h:
```diff
@@ -917,6 +924,8 @@ struct RGWBucketInfo {
 
   std::map<std::string, uint32_t> mdsearch_config;
 
+  rgw::cksum::Type cksum_type = rgw::cksum::Type::none;
```
this is an artifact from my original PR which stored a checksum preference with the bucket info;  Maybe this should go away...

cbodley commented:
i revisited the discussion on @imtzw's original pr and was reminded of this one: https://github.com/ceph/ceph/pull/49986#issuecomment-1433578439

> from https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#large-object-checksums
>
>> If you've enabled additional checksum values for your multipart object, Amazon S3 calculates the checksum for each individual part by using the specified checksum algorithm. The checksum for the completed object is calculated in the same way that Amazon S3 calculates the MD5 digest for the multipart upload. You can use this checksum to verify the integrity of the object.
>
> the existing logic for ETag/MD5 is here: https://github.com/ceph/ceph/blob/d8dcffa/src/rgw/driver/rados/rgw_sal_rados.cc#L2700-L2707

it doesn't look like this multipart behavior was implemented here either?

cbodley commented on src/rgw/rgw_xxh_digest.h:
```diff
@@ -0,0 +1,53 @@
+#include <stdio.h>
+#include "rgw_crc_digest.h"
+
+#define XXH_INLINE_ALL 1 /* required for streaming variants */
+#include "xxhash.h"
```
i'm seeing compilation failures here and that define looks related? with gcc 13.2.1
```
In file included from /home/cbodley/ceph/src/rgw/rgw_xxh_digest.h:23,
                 from /home/cbodley/ceph/src/rgw/rgw_cksum.h:30,
                 from /home/cbodley/ceph/src/rgw/rgw_cksum_pipe.h:21,
		 from /home/cbodley/ceph/src/rgw/rgw_cksum_pipe.cc:16:
In function ‘void XXH3_accumulate(xxh_u64*, const xxh_u8*, const xxh_u8*, size_t, XXH3_f_accumulate_512)’,
    inlined from ‘void XXH3_hashLong_internal_loop(xxh_u64*, const xxh_u8*, size_t, const xxh_u8*, size_t, XXH3_f_accumulate_512, XXH3_f_scrambleAcc)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:4391:24,
    inlined from ‘XXH64_hash_t XXH3_hashLong_64b_internal(const void*, size_t, const void*, size_t, XXH3_f_accumulate_512, XXH3_f_scrambleAcc)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:4454:32,
    inlined from ‘XXH64_hash_t XXH3_hashLong_64b_withSecret(const void*, size_t, XXH64_hash_t, const xxh_u8*, size_t)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:4474:38,
    inlined from ‘XXH64_hash_t XXH3_64bits_internal(const void*, size_t, XXH64_hash_t, const void*, size_t, XXH3_hashLong64_f)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:4555:22,
    inlined from ‘XXH64_hash_t XXH_INLINE_XXH3_64bits_withSecret(const void*, size_t, const void*, size_t)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:4571:32:
/home/cbodley/ceph/src/xxHash/xxhash.h:3932:1: error: inlining failed in call to ‘always_inline’ ‘void XXH3_accumulate_512_sse2(void*, const void*, const void*)’: function not considered for inlining
 3932 | XXH3_accumulate_512_sse2( void* XXH_RESTRICT acc,
      | ^~~~~~~~~~~~~~~~~~~~~~~~
```

mattbenjamin replied:
rrrr

mattbenjamin replied:
you need the define to expose the required defns;  I'll work on it

mattbenjamin replied:
this seems to be a known bug [1] ; gcc-13.2 compile completed successfully after I pulled xxHash up to .0.8.2
[1] https://github.com/Cyan4973/xxHash/issues/800

cbodley replied:
works for me now 👍

cbodley replied:
i saw that epel9 has xxhash 0.8.2 but ubuntu jammy only has 0.8.1. so close to being able to drop the submodule 🤷

cbodley commented on src/rgw/rgw_basic_types.h:
```diff
@@ -31,6 +31,7 @@
 #include "rgw_user_types.h"
 #include "rgw_bucket_types.h"
 #include "rgw_obj_types.h"
+#include "rgw_cksum.h"
```
rgw_cksum.h is all header-only, so this pulls a bunch of header dependencies into rgw_basic_types.h that aren't needed here. all that `RGWUploadPartInfo` really needs is a struct containing the Type enum, the digest value array, and their encodings

could we separate out a rgw_cksum_types.h that only has the essentials?

cbodley replied:
i remember seeing some changes related to this, but they've since disappeared. what happened to rgw_cksum_types.h?

mattbenjamin replied:
oh, yeah.  I thought that was still present.  I'll look through my snapshot branches

mattbenjamin replied:
(I rebased a bunch of times)

cbodley commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -481,6 +488,20 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
     }
 
+    if (checksum_mode) {
+      if (auto i = attrs.find("RGW_ATTR_AMZ_CHECKSUM"); i != attrs.end()) {
```
`attrs.find(RGW_ATTR_AMZ_CKSUM)`

cbodley replied:
you added a define for this xattr name, that's what we need to use here:
```c++
#define RGW_ATTR_AMZ_CKSUM      RGW_ATTR_PREFIX "x-amz-content-checksum"
```

mattbenjamin replied:
(outdated)

cbodley commented:
[PutObject](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html) is missing response headers `x-amz-checksum-crc32` etc

[CreateMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html) missing `x-amz-checksum-algorithm`  as request header and response header

[CompleteMultipartUpload](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) is missing:

- read the `<Part>` checksums out of the `<CompleteMultipartUpload>` document and compare against stored part checksums
- require checksum algorithm to match CreateMultipartUpload and all part uploads
- calculate the composite checksum of parts
- parse request headers `x-amz-checksum-crc32` etc. and, if present, compare against the composite checksum
- store composite checksum in an object xattr for Head/GetObject (`RGW_ATTR_AMZ_CKSUM` or something else? needs to include the extra parts-count string at the end)
- write composite checksum in response header ie `x-amz-checksum-crc32`

cbodley commented:
i pushed https://github.com/cbodley/ceph/commits/wip-rgw-cksum, which is a rebase of this branch on top of the latest in https://github.com/ceph/ceph/pull/54856/, plus 3 commits which get the non-multipart test cases in https://github.com/ceph/s3-tests/pull/540 passing. i still can't get xxhash to compile so i commented it out of rgw_cksum.h, but didn't include those changes

mattbenjamin commented on src/rgw/rgw_cksum.h:
```diff
@@ -0,0 +1,296 @@
+
+    void decode(bufferlist::const_iterator& p) {
+      DECODE_START(1, p);
+      uint16_t tt;
+      decode(tt, p);
+      type = cksum::Type(tt);
```
this was needed at one point, iirc

mattbenjamin replied:
if the right thing happens automatically now, will remove this

mattbenjamin replied:
ok, reviewed this;  I'm I need a temporary to recover the max_digest_size anyway, unless you think this is confusing, can I leave this as is?

cbodley replied:
sure. this is problably going to need further changes to support the composite checksums for multipart which add the extra parts-count suffix at the end

mattbenjamin commented on src/rgw/rgw_op.cc:
```diff
@@ -4497,6 +4506,28 @@ void RGWPutObj::execute(optional_yield y)
+    auto cksum_verify =
+      cksum_filter->verify(*s->info.env); // valid or no supplied cksum
+    auto& cksum = get<1>(cksum_verify);
+    if (std::get<0>(cksum_verify)) {
+      buffer::list cksum_bl;
+      cksum_bl.append(cksum.to_string());
+      emplace_attr(RGW_ATTR_CKSUM, std::move(cksum_bl));
```
I will recheck to_string, that's probably correct

mattbenjamin commented on src/rgw/rgw_op.cc:
```diff
@@ -4497,6 +4506,28 @@ void RGWPutObj::execute(optional_yield y)
+    } else {
+      /* content checksum mismatch */
+      const auto& hdr = cksum_filter->header();
+      ldpp_dout(this, 4)
+	<< fmt::format("{} content checksum mismatch", hdr.second)
+	<< fmt::format("\n\tcalculated={} != \n\texpected={}",
+		       cksum.to_armor(),
+		       cksum_filter->expected(*s->info.env))
+	<< dendl;
```
don't know how much that matters, I prefer fmt syntax almost always

mattbenjamin commented on src/rgw/rgw_cksum_pipe.h:
```diff
@@ -0,0 +1,93 @@
+  namespace cksum = rgw::cksum;
+  using cksum_hdr_t = std::pair<const char*, const char*>;
+  
+  // PutObj filter for streaming checksums
+  class RGWPutObj_Cksum : public rgw::putobj::Pipe {
```
I have to say, the name was most enjoyable

cbodley commented:
> it's possible that the blake3 and xx3 mechanisms should be skipped for now; I really wish to have them for future--but for now, there'd be no way to access them; it does seem like we could switch our xxh64 callers to xx3, almost certainly?

agree that we should focus on the sdk-supported algorithms for now

i do think we can use our [boto extension strategy](https://github.com/ceph/ceph/tree/main/examples/rgw/boto3) to teach it about the new checksum algorithms. that should work for the low-level operations like put_object() etc. where the caller passes in their own computed checksum. we just won't be able to get the higher-level sdk operations like upload_file() to compute those checksums for us

yuriw commented:
@mattbenjamin pls rebase

mattbenjamin commented:
@yuriw rebasing!

mattbenjamin commented:
@cbodley xxHash is pulled up again, resolves the always-inline issue;  more to do, obviously;  noticed that one of the recent qat changes (maybe just enabling it?) causes an error compiling qat_internal.h with clang 16; not going to worry about that right now

cbodley commented:
> noticed that one of the recent qat changes (maybe just enabling it?) causes an error compiling qat_internal.h with clang 16; not going to worry about that right now

thanks, i reproduced that with clang 17 and opened https://tracker.ceph.com/issues/64472 and https://github.com/intel/QATzip/issues/111

mattbenjamin commented on src/rgw/rgw_cksum_pipe.cc:
```diff
@@ -0,0 +1,84 @@
+    return 0;
+  }
+
+  RGWPutObj_Cksum::~RGWPutObj_Cksum()
+  {
+    if ((_state > State::START) &&
+	(_state < State::FINAL))
+      (void) finalize();
+  }
```
I had a kind of though process here, but sure, will drop this

cbodley commented:
i removed the squid milestone from this pr because it's confusing the squid backport/testing process

dang commented on src/rgw/rgw_op.cc:
```diff
@@ -4517,7 +4519,6 @@ void RGWPutObj::execute(optional_yield y)
     cksum = get<1>(cksum_verify);
     if (std::get<0>(cksum_verify)) {
       buffer::list cksum_bl;
-      //cksum_bl.append(cksum->to_string());
```
This may make no difference, but it did jump out at me as an odd change.

dang commented on src/rgw/rgw_op.cc:
```diff
@@ -6359,9 +6364,103 @@ void RGWCompleteMultipart::pre_exec()
+     7. verify -- if invalid, fail */
+
+    /* rgw_sal.h says that list_parts is called for the side effect of loading
+     * the parts of an upload into "cache"--the api is strange and truncated
```
The "cache" is just a map of MultipartPart objects.  Resuming a truncated list_parts() call on the same MultipartUpload object continues to fill it's map, so a complete list_parts() run (i.e. until !truncated) should have a complete map of parts.

mattbenjamin replied:
but if we hit the truncated case, and then re-enter list_parts(), what happens to the cached info elements?

dang commented on src/rgw/rgw_op.cc:
```diff
@@ -6359,9 +6364,103 @@ void RGWCompleteMultipart::pre_exec()
+  rgw::cksum::Cksum* prev_cksum{nullptr};
+
+  /* XXX as presently, it appears that if 'truncated' is passed as 'true' to
+   * list_parts, its value is never updated */
```
Odd.  It looks like 'truncated' is always set and never read in RadosStore.

mattbenjamin replied:
I also missed resetting it to false in the truncated case, just fixed that

dang reviewed (commented):
Everything looks good in the multipart upload commit.  Nothing jumps out at me as being broken.

dang commented on src/rgw/rgw_op.cc:
```diff
@@ -4793,16 +4810,41 @@ void RGWPostObj::execute(optional_yield y)
     }
 
-    /* TODO:  implement POST checksums */
+    if (false && cksum_filter) {
```
This will never run?

mattbenjamin replied:
looks like I intended to come back to that--thanks for spotting it

cbodley commented on src/rgw/rgw_cksum_pipe.cc:
```diff
@@ -0,0 +1,91 @@
+    return std::unique_ptr<RGWPutObj_Cksum>();
+  }
+
+  int RGWPutObj_Cksum::process(ceph::buffer::list &&data, uint64_t logical_offset)
+  {
+    for (const auto& ptr : data.buffers()) {
+      _digest->Update(reinterpret_cast<const unsigned char*>(ptr.c_str()),
+                      ptr.length());
+    }
+    return 0;
+  }
```
this filter wraps a `next` filter as part of the putobj pipeline. in addition to using this input `data` to calculate checksums, we also need to pass it on to the next filter:
```c++
 return Pipe::process(std::move(data), logical_offset);
```

mattbenjamin replied:
ah

mattbenjamin replied:
for shame!

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -4493,6 +4513,34 @@ void RGWPutObj::execute(optional_yield y)
+    cksum = get<1>(cksum_verify);
+    if (std::get<0>(cksum_verify)) {
+      buffer::list cksum_bl;
+      ldpp_dout(this, 16)
+	<< fmt::format("{} content checksum verified ", hdr.second)
+	<< fmt::format("\n\tcalculated={} == \n\texpected=  {}",
+		       cksum->to_armor(),
+		       cksum_filter->expected(*s->info.env))
+	<< dendl;
```
this dout+format pattern is showing up a lot here, and i argued against it in https://github.com/ceph/ceph/pull/56008#pullrequestreview-1944064024

both of these format() calls are going to allocate memory because the result is longer than the short string optimization. is there some reason to expect that this is still better than the style used elsewhere in ceph?
```c++
ldpp_dout(this, 16) << hdr.second << " content checksum verified\n\t"
    "calculated=" << cksum->to_armor() << " == \n\t"
    "expected=  " << cksum_filter->expected(*s->info.env))
    << dendl;
```

cbodley replied:
> the potential difference iirc is that we expect each of these '<<' operator invocations to take locks

that's not the case, no. you might be thinking about the locale-related locking in std::ostream's constructor? in https://github.com/ceph/ceph/pull/19100 (with additional discussion in https://github.com/ceph/ceph/pull/17938) we made dout's ostream thread-local so it's only constructed once per thread and reused for each call

mattbenjamin replied:
I just don't think use of format to reduce << spam is an antipattern.

cbodley replied:
i hacked together a set of dout macros for this in https://github.com/ceph/ceph/pull/58229

mattbenjamin reviewed (commented):
for some reason github says I'm reviewing

mattbenjamin commented:
@cbodley @dang this is passing local jcksum and s3-tests (branch: https://github.com/ceph/s3-tests/pull/564)

mattbenjamin commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -3993,6 +4031,9 @@ void RGWInitMultipart_ObjStore_S3::send_response()
     dump_time_header(s, "x-amz-abort-date", abort_date);
     dump_header_if_nonempty(s, "x-amz-abort-rule-id", rule_id);
   }
+  if (auto i = attrs.find(RGW_ATTR_CKSUM_ALGORITHM); i != attrs.end()) {
```
ah, got it, it's no longer needed

mattbenjamin commented on src/rgw/rgw_op.cc:
```diff
@@ -4358,15 +4366,27 @@ void RGWPutObj::execute(optional_yield y)
+      cksum_filter =
+	rgw::putobj::RGWPutObj_Cksum::Factory(filter, *s->info.env);
+    } catch (const rgw::io::Exception& e) {
+      op_ret = e.code().value();
```
I think I missed this one, will check

mattbenjamin commented:
> i pushed https://github.com/cbodley/ceph/commits/wip-rgw-cksum, which is a rebase of this branch on top of the latest in #54856, plus 3 commits which get the non-multipart test cases in [ceph/s3-tests#540](https://github.com/ceph/s3-tests/pull/540) passing. i still can't get xxhash to compile so i commented it out of rgw_cksum.h, but didn't include those changes

ok, that is ideal.  I'll rebase on it when you get some more acks?

cbodley commented on src/rgw/CMakeLists.txt:
```diff
@@ -57,6 +84,7 @@ set(librgw_common_srcs
   services/svc_user_rados.cc
   services/svc_zone.cc
   services/svc_zone_utils.cc
+  ${blake3_srcs}
```
it doesn't seem right to add the blake3 sources directly to the rgw_common target. you've exposed the Blake3 digest in common/ceph_crypto.h but it wouldn't be useable outside of rgw

we should either link it into ceph-common, or split that out of common/ceph_crypto.h into a separate library with its own header, and have rgw depend on that

p.s. blake3 comes with its own [CMakeLists.txt](https://github.com/BLAKE3-team/BLAKE3/blob/master/c/CMakeLists.txt) that i'd hope we could use directly by putting `add_subdirectory(BLAKE3/c)` in our `src/CMakeLists.txt` and linking to its `BLAKE3::blake3` target

that should also set up the correct include directory so that we can `#include <blake3.h>` instead of `"BLAKE3/c/blake3.h"` which won't work if we ever switch from submodule to packages. on that note, i found blake3-devel packages in fedora/epel, but ubuntu/debian only package the rust library librust-blake3-dev

mattbenjamin replied:
I don't mind using their submodule, if it works out of the box.  I honestly think that these changes, could be in future commits, e.g., when another component wants to use blake3 or someone decides that platform integration with this blake3 implementation is now possible--it definitely would be a waste of effort now.

mattbenjamin replied:
(i.e., to worry about platform packages)

cbodley commented:
> > i pushed https://github.com/cbodley/ceph/commits/wip-rgw-cksum, which is a rebase of this branch on top of the latest in #54856, plus 3 commits which get the non-multipart test cases in [ceph/s3-tests#540](https://github.com/ceph/s3-tests/pull/540) passing. i still can't get xxhash to compile so i commented it out of rgw_cksum.h, but didn't include those changes
> 
> ok, that is ideal. I'll rebase on it when you get some more acks?

is my branch still relevant? i thought you got those tests passing already

mattbenjamin commented:
> > > i pushed https://github.com/cbodley/ceph/commits/wip-rgw-cksum, which is a rebase of this branch on top of the latest in #54856, plus 3 commits which get the non-multipart test cases in [ceph/s3-tests#540](https://github.com/ceph/s3-tests/pull/540) passing. i still can't get xxhash to compile so i commented it out of rgw_cksum.h, but didn't include those changes
> > 
> > 
> > ok, that is ideal. I'll rebase on it when you get some more acks?
> 
> is my branch still relevant? i thought you got those tests passing already

when I wrote this, I thought we'd absorb your dout_fmt branch;  just ignore

cbodley commented on src/rgw/driver/rados/rgw_rados.cc:
```diff
@@ -6736,6 +6739,12 @@ int RGWRados::Object::Read::prepare(optional_yield y, const DoutPrefixProvider *
+  if (manifest /* params.parts_count */) {
+      RGWObjManifest::obj_iterator end = manifest->obj_end(dpp);
+      auto cur_part_id = end.get_cur_part_id();
+      params.parts_count = (cur_part_id == 1) ? 1 : cur_part_id - 1;;
```
the comment now reads `If the object is multipart, the total number of multipart parts is assigned to this output parameter.`, but it looks like this sets a `parts_count` for non-multipart uploads too. can we find a way to leave `parts_count` as an empty optional when it's not a multipart manifest?

mattbenjamin replied:
is that really helpful?

mattbenjamin replied:
maybe it's easy

mattbenjamin replied:
the general reason that this logic would have a manifest, is that this is a multipart upload.  I don't think this really needs to be changed.

cbodley replied:
i'm concerned that this is going to cause Get/HeadObject requests to start returning x-amz-mp-parts-count headers for non-multipart uploads where they didn't before, and that this doesn't do what the comment says it does

mattbenjamin replied:
it appears that RWObjManifest has a RGWObjTier which has a boolean member "is_multipart_upload".  I do not know if it would be safe or correct to add that to the conditional.

mattbenjamin replied:
well the comment can be changed.  but afaik, the only potential issue would be with non-multipart objects with a manifest.  do I have a way to test for that?  doing so should resolve that issue

cbodley replied:
`end.get_cur_part_id()` returns 0 for non-multipart manifests

mattbenjamin replied:
ok

mattbenjamin replied:
done

cbodley commented on src/rgw/rgw_rest_s3.cc:
```diff
@@ -473,7 +482,7 @@ int RGWGetObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t bl_ofs,
   }
 
-  if (multipart_parts_count) {
+  if (multipart_parts_count && multipart_parts_count > 0) {
```
since `multipart_parts_count` is a `std::optional<int>`, shouldn't these comparisons be `*multipart_parts_count > 0`? it's also not clear how it would end up with a 0 value given the code in rgw_rados.cc:
```c++
params.parts_count = (cur_part_id == 1) ? 1 : cur_part_id - 1;;
```

mattbenjamin replied:
that's true but this is defensive programming that tracks the structure of the variable

mattbenjamin replied:
yes, I misread

cbodley commented on src/rgw/rgw_xxh_digest.h:
```diff
@@ -0,0 +1,53 @@
+      }
+      memcpy((char*) digest, &final, sizeof(final));
+    }
+  }; /* Crc32c */
```
typo: this is the of `class XXH3`

mattbenjamin replied:
that's already fixed in the last push

cbodley reviewed (commented):
rgw_cksum_digest.h is missing from this branch:
```
src/rgw/rgw_cksum_pipe.h:23:10: fatal error: rgw_cksum_digest.h: No such file or directory
23 | #include "rgw_cksum_digest.h"
   |          ^~~~~~~~~~~~~~~~~~~~
```

cbodley commented on src/rgw/rgw_op.cc:
```diff
@@ -6329,14 +6529,32 @@ void RGWCompleteMultipart::execute(optional_yield y)
+		(!!upload) ? to_string(upload->cksum_type) : 0)
+		<< dendl;
+
+  rgw_placement_rule* dest_placement;
+  op_ret = upload->get_info(this, s->yield, &dest_placement);
+  if (op_ret < 0) {
+    /* XXX this fails consistently when !checksum */
+    ldpp_dout(this, 0) <<
+      "WARNING: MultipartUpload::get_info() for placement failed "
+		       << "ret=" << op_ret << dendl;
+    if (upload->cksum_type != rgw::cksum::Type::none) {
+      op_ret =  -ERR_INTERNAL_ERROR;
```
ok, i guess we need this for `upload->cksum_type` below? we do need to know that so we can reject the request before calling `upload->complete()`. does `upload->get_info()` cache everything we'd need later in `RadosMultipartUpload::complete()`? if so i hope we could remove its own call to `RadosMultipartUpload::get_info()`

mattbenjamin replied:
I don't know.  I do not want to wait to merge checksums based on a theory that we should change any of the internals of the RadosMultipartUpload

mattbenjamin replied:
I will experiment with this tonight;  if it works as expected, we don't need to call get_info twice (but I'm not sure that adds actual overhead)

cbodley replied:
`RadosMultipartUpload::get_info()` calls `read_op->prepare()` and `read_op->read()`. because of `set_prefetch_data()`, the `read()` call reads from the prefetched buffer instead of going back to the osd

if a repeated call to `RadosMultipartUpload::get_info()` used the same `sal::Object` for meta_obj, then both `prepare()` and `read()` could be satisfied from its `RGWObjectCtx` cache

however, `get_info()` uses `get_meta_obj()` to create its own local `sal::Object` handle. because they don't share a `RGWObjectCtx`, i expect that duplicated calls do add an extra round trip to the osd

mattbenjamin replied:
I believed that the call was required prior to invoking list_parts(), which is required to validate the checksums in the upload

mattbenjamin replied:
we also consult the checksum type of the original upload, but I think that's more a consistency check

mattbenjamin replied:
ok, so per webex discussion, our strategy to address this is to take daniel's change to cache attributes in RadosMultipartUpload.   thank you!

mattbenjamin commented on src/test/rgw/test_rgw_cksum.cc:
```diff
@@ -0,0 +1,386 @@
+  ASSERT_EQ(ck2.to_armor(), ck3.first.to_armor());
+}
+
+TEST(RGWCksum, Output)
+{
+  auto o_mode = std::ios::out|std::ios::trunc;
+  std::ofstream of;
+  of.open("/tmp/lorem", o_mode);
+  of << lorem;
+  of.close();
+
+  of.open("/tmp/dolor", o_mode);
+  of << dolor;
+  of.close();
+}
```
this was useful for interacting with the output from outside the test.   is this really a problem?  e.g., you've started running this on windows

mattbenjamin commented on src/rgw/rgw_op.cc:
```diff
@@ -6387,17 +6613,47 @@ void RGWCompleteMultipart::execute(optional_yield y)
   }
-  s->object->set_attrs(meta_obj->get_attrs());
+
+  auto target_attrs = meta_obj->get_attrs();
```
ack

mattbenjamin replied:
it cannot be a const ref, due to emplace at l. 6638.  afaik, it could be a non-const ref

mattbenjamin commented:
> rgw_cksum_digest.h is missing from this branch:
> 
> ```
> src/rgw/rgw_cksum_pipe.h:23:10: fatal error: rgw_cksum_digest.h: No such file or directory
> 23 | #include "rgw_cksum_digest.h"
>    |          ^~~~~~~~~~~~~~~~~~~~
> ```

I updated in slack regarding that.  I've rebased the branch and fixed this issue

mattbenjamin commented on src/rgw/rgw_cksum.h:
```diff
@@ -0,0 +1,231 @@
+#include <iterator>
+#include <boost/algorithm/string.hpp>
+#include "fmt/format.h"
+#include "common/ceph_crypto.h"
+#include "common/armor.h"
+#include "rgw_blake3_digest.h"
+#include "rgw_crc_digest.h"
+#include "rgw_xxh_digest.h"
```
yes, that's desired

cbodley commented:
trying to compile this locally with gcc 14.1.1, i'm getting the following errors from xxhash:
```
In file included from /home/cbodley/ceph/src/rgw/rgw_blake3_digest.h:23,
                 from /home/cbodley/ceph/src/rgw/rgw_cksum_digest.h:19:
In function ‘void XXH3_hashLong_internal_loop(xxh_u64*, const xxh_u8*, size_t, const xxh_u8*, size_t, XXH3_f_accumulate, XXH3_f_scrambleAcc)’,
    inlined from ‘XXH64_hash_t XXH3_hashLong_64b_internal(const void*, size_t, const void*, size_t, XXH3_f_accumulate, XXH3_f_scrambleAcc)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:5607:32,
    inlined from ‘XXH64_hash_t XXH3_hashLong_64b_withSeed_internal(const void*, size_t, XXH64_hash_t, XXH3_f_accumulate, XXH3_f_scrambleAcc, XXH3_f_initCustomSecret)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:5666:42,
    inlined from ‘XXH64_hash_t XXH3_hashLong_64b_withSeed(const void*, size_t, XXH64_hash_t, const xxh_u8*, size_t)’ at /home/cbodley/ceph/src/xxHash/xxhash.h:5685:47:
/home/cbodley/ceph/src/xxHash/xxhash.h:4822:1: error: inlining failed in call to ‘always_inline’ ‘void XXH3_scrambleAcc_sse2(void*, const void*)’: function not considered for inlining
 4822 | XXH3_scrambleAcc_sse2(void* XXH_RESTRICT acc, const void* XXH_RESTRICT secret)
      | ^~~~~~~~~~~~~~~~~~~~~
/home/cbodley/ceph/src/xxHash/xxhash.h:5545:19: note: called from here
 5545 |         f_scramble(acc, secret + secretSize - XXH_STRIPE_LEN);
      |         ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

cbodley commented on src/rgw/rgw_blake3_digest.h:
```diff
@@ -0,0 +1,47 @@
+#include <stdio.h>
+#include "BLAKE3/c/blake3.h"
+
+#define XXH_INLINE_ALL 1 /* required for streaming variants */
+#include "xxhash.h"
```
rgw_blake3_digest.h shouldn't need this xxhash.h include

cbodley reviewed (approved):
looks great as long as ci builds work. are you planning to squash commits?

mattbenjamin commented:
> looks great as long as ci builds work. are you planning to squash commits?

yes, I will squash some more

cbodley commented:
> edit: this was happening before, and it went away when you bumped xxhash to 0.8.2. my submodule is updated, but i'll try a clean rebuild (no luck after rebuild)

i was able to compile successfully after adding `-DCMAKE_CXX_FLAGS_DEBUG=-g\ -DXXH_NO_INLINE_HINTS=1` to `-DCMAKE_BUILD_TYPE=Debug` in my cmake command line. i'll think about how best to get our CMakeLists.txt to handle this for us in debug builds

mattbenjamin commented:
> > edit: this was happening before, and it went away when you bumped xxhash to 0.8.2. my submodule is updated, but i'll try a clean rebuild (no luck after rebuild)
> 
> i was able to compile successfully after adding `-DCMAKE_CXX_FLAGS_DEBUG=-g\ -DXXH_NO_INLINE_HINTS=1` to `-DCMAKE_BUILD_TYPE=Debug` in my cmake command line. i'll think about how best to get our CMakeLists.txt to handle this for us in debug builds

great, thanks casey

mattbenjamin commented:
{
	  auto mtime_as_timet = ceph::real_clock::to_time_t(mtime);
	  auto days = rule_expiration.get_days();
	  auto raw_span = double(rule_expiration.get_days()) * secs_in_a_day - ceph::real_clock::to_time_t(mtime)%(secs_in_a_day) + secs_in_a_day;
	  auto span = make_timespan(raw_span);
	  auto expire_at = mtime + span;
	  ldpp_dout_fmt(dpp, 16,
			"mystere timey stuff: mtime: {} mtime_as_timet: {} days: {} raw_span: {} span: {} computed expire_at: {} ",
			mtime, mtime_as_timet, days, raw_span, span, expire_at);
	  ldpp_dout(dpp, 16) << "timey test" << dendl;
	  
	}

cbodley commented:
from 'make check':
```
src/test/rgw/test_rgw_cksum.cc:40:15: error: unused variable 't0' [-Werror,-Wunused-variable]
  cksum::Type t0 = cksum::Type::none;
              ^
1 error generated.
```

mattbenjamin commented:
generic rgw suite run I think passed:
https://pulpito.ceph.com/mbenjamin-2024-06-29_19:25:01-rgw-wip-rgw-cksum-distro-default-smithi/

2nd rgw suite run with new checksum tests in progress:
https://pulpito.ceph.com/mbenjamin-2024-07-01_15:11:31-rgw-wip-rgw-cksum-distro-default-smithi/

mattbenjamin commented:
rerun:
https://pulpito.ceph.com/mbenjamin-2024-07-01_18:24:14-rgw-wip-rgw-cksum-distro-default-smithi/

cbodley commented:
rgw/dbstore:
> FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_sha256 - ...
FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_3parts - ...

a lot of multipart test cases are filtered out using `@pytest.mark.fails_on_dbstore`, these tests probably need the same

rgw/lifecycle:
> FAILED s3tests_boto3/functional/test_s3.py::test_lifecycle_expiration_size_gt

i can't tell why that one failed

mattbenjamin commented:
> rerun: https://pulpito.ceph.com/mbenjamin-2024-07-01_18:24:14-rgw-wip-rgw-cksum-distro-default-smithi/

@cbodley--does this look like a pass?

mattbenjamin commented:
> rgw/dbstore:
> 
> > FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_sha256 - ...
> > FAILED s3tests_boto3/functional/test_s3.py::test_multipart_checksum_3parts - ...
> 
> a lot of multipart test cases are filtered out using `@pytest.mark.fails_on_dbstore`, these tests probably need the same
> 
> rgw/lifecycle:
> 
> > FAILED s3tests_boto3/functional/test_s3.py::test_lifecycle_expiration_size_gt
> 
> i can't tell why that one failed

I've been seeing some unexpected lc failures--I saw this in a scenario where it looked like my test system went into screen lock;  more recently, I saw consistent failures during specific times of the day (but not in this test).

mattbenjamin commented:
rerun 3 (rerun 2 had wrong baseline):
https://pulpito.ceph.com/mbenjamin-2024-07-03_13:56:10-rgw-wip-rgw-cksum-distro-default-smithi/

mattbenjamin commented:
jenkins retest this please

cbodley commented on .gitmodules:
```diff
@@ -75,3 +75,10 @@
 [submodule "src/qatzip"]
 	path = src/qatzip
 	url = https://github.com/intel/qatzip.git
+[submodule "src/boost_redis"]
+	path = src/boost_redis
+	url = https://github.com/boostorg/redis.git
```
this was added back when you resolved the merge conflict. i opened https://github.com/ceph/ceph/pull/58428 to remove it again

cbodley commented:
> > i'll think about how best to get our CMakeLists.txt to handle this for us in debug builds
> 
> @mattbenjamin this what i came up with:
> 
> ```cmake
> # work around https://github.com/Cyan4973/xxHash/issues/943 for debug builds
> target_compile_definitions(rgw_common PUBLIC
>   $<$<CONFIG:Debug>:XXH_NO_INLINE_HINTS=1>
>   $<$<CONFIG:RelWithDebInfo>:XXH_NO_INLINE_HINTS=1>)
> ```
> 
> would you like me to push a commit for that here? it's currently in [ea9f13d](https://github.com/ceph/ceph/commit/ea9f13d9f1ebaf2397e2e2fb2d787c5c6fc5bde8) as part of #58310, which goes on to add support for blake3-devel and xxhash-devel system packages when they exist

we missed this part, so main fails to build in debug mode. i pushed my commit to a new branch and opened https://github.com/ceph/ceph/pull/58429 to fix

rkhudov commented:
@cbodley @mattbenjamin 
I have multiple issues after merging this PR:

1. 
```
src/rgw/rgw_xxh_digest.h:29:5: error: 'XXH3_state_t' does not name a type; did you mean 'XXH32_state_t'?
XXH3_state_t s;
```

2.
```
src/rgw/rgw_xxh_digest.h:35:23: error: 's' was not declared in this scope
XXH3_INITSTATE(&s);
```

3.
```
src/rgw/rgw_xxh_digest.h:39:41: error: 's' was not declared in this scope
void Restart() { XXH3_64bits_reset(&s); }
```

4.
```
src/rgw/rgw_xxh_digest.h:39:22: error: 'XXH3_64bits_reset' was not declared in this scope; did you mean 'XXH64_reset'?
 void Restart() { XXH3_64bits_reset(&s); }
```
